**START**

Load in pandas and sqlalchemy's create_engine & text
Then connect to Postgres and create the engine

In [1]:
from sqlalchemy import create_engine, text
import pandas as pd
#import collections as coll
import geopandas as gpd
#import matplotlib.pyplot as plt
import folium
from folium.plugins import MarkerCluster
database_name = 'scooters'
connection_string = f"postgresql://postgres:postgres@localhost:5433/{database_name}"
engine = create_engine(connection_string)

**EDA**

In [2]:
count_all_rows = '''
(SELECT
    'scooters' table,
    COUNT(*)
FROM scooters)
UNION
(SELECT
    'trips' table,
    COUNT(*)
FROM trips);
'''

with engine.connect() as connection:
    counts = pd.read_sql(text(count_all_rows), con = connection)

counts

table     count
0  scooters  73414043
1     trips    565522

In [3]:
find_nulls_scooters = '''
SELECT *
FROM scooters
WHERE NOT(scooters IS NOT NULL);
'''

with engine.connect() as connection:
    nulls_scooters = pd.read_sql(text(find_nulls_scooters), con = connection)

find_nulls_trips = '''
SELECT *
FROM trips
WHERE NOT(trips IS NOT NULL);
'''

with engine.connect() as connection:
    nulls_trips = pd.read_sql(text(find_nulls_trips), con = connection)

nulls_scooters.info()
nulls_trips.info()

nulls_scooters

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 770 entries, 0 to 769
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   pubdatetime  770 non-null    datetime64[ns]
 1   latitude     770 non-null    float64       
 2   longitude    770 non-null    float64       
 3   sumdid       770 non-null    object        
 4   sumdtype     770 non-null    object        
 5   chargelevel  0 non-null      object        
 6   sumdgroup    770 non-null    object        
 7   costpermin   770 non-null    float64       
 8   companyname  770 non-null    object        
dtypes: datetime64[ns](1), float64(3), object(5)
memory usage: 54.3+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 0 entries
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   pubtimestamp    0 non-null      object
 1   companyname     0 non-null      object
 2   triprecordn

pubdatetime   latitude  longitude  \
0   2019-06-15 20:15:30  36.158659 -86.780577   
1   2019-06-15 20:20:30  36.158659 -86.780577   
2   2019-06-15 20:25:30  36.158659 -86.780577   
3   2019-06-16 22:41:30  36.166987 -86.759158   
4   2019-06-16 22:46:31  36.166991 -86.759138   
..                  ...        ...        ...   
765 2019-06-14 18:34:33  36.150252 -86.813099   
766 2019-06-14 19:04:34  36.150271 -86.813111   
767 2019-06-14 19:09:34  36.150271 -86.813111   
768 2019-06-14 19:14:34  36.150274 -86.813108   
769 2019-06-15 20:10:30  36.158659 -86.780577   

                                           sumdid sumdtype chargelevel  \
0    Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
1    Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
2    Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
3    Powered-9b20f3b4-4a3d-ab88-a230-ca90d32c8fc2  Powered        None   
4    Powered-9b20f3b4-4a3d-ab88-a230-ca90d32c8fc2  Powered        None   
..                                            ...      ...         ...   
765  Powered-72a04621-3a01-05f2-3c9b-abae6d7387ce  Powered        None   
766  Powered-72a04621-3a01-05f2-3c9b-abae6d7387ce  Powered        None   
767  Powered-72a04621-3a01-05f2-3c9b-abae6d7387ce  Powered        None   
768  Powered-72a04621-3a01-05f2-3c9b-abae6d7387ce  Powered        None   
769  Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   

    sumdgroup  costpermin companyname  
0     Scooter        0.15        Bolt  
1     Scooter        0.15        Bolt  
2     Scooter        0.15        Bolt  
3     Scooter        0.15        Bolt  
4     Scooter        0.15        Bolt  
..        ...         ...         ...  
765   Scooter        0.15        Bolt  
766   Scooter        0.15        Bolt  
767   Scooter        0.15        Bolt  
768   Scooter        0.15        Bolt  
769   Scooter        0.15        Bolt  

[770 rows x 9 columns]

While the .info() part of the output is a bit counter-intuitave, it shows that there are 770 null values in the scooters table and none in the trips table.

All of the null values are in the chargelevel column, and looking at the full output, they all belong to Bolt and Spin.

In [4]:
date_range = '''
(SELECT
    'scooters' table,
    MIN(pubdatetime) begin,
    MAX(pubdatetime) end
FROM scooters)
UNION
(SELECT
    'trips' table,
    MIN(pubtimestamp) begin,
    MAX(pubtimestamp) end
FROM trips);
'''

with engine.connect() as connection:
    dates = pd.read_sql(text(date_range), con = connection)

dates

table                   begin                 end
0  scooters 2019-05-01 00:01:41.247 2019-07-31 23:59:57
1     trips 2019-05-01 00:00:55.423 2019-08-01 07:04:00

In [5]:
aug_first = '''
SELECT *
FROM trips
WHERE enddate > '2019-07-31'
LIMIT 100;
'''

with engine.connect() as connection:
    late = pd.read_sql(text(aug_first), con = connection)

late

pubtimestamp companyname triprecordnum  \
0  2019-08-01 00:01:02.110        Lyft          LFT2   
1  2019-08-01 00:00:06.593        Bird       BRD1816   
2  2019-08-01 00:00:06.593        Bird       BRD1819   
3  2019-08-01 00:00:06.593        Bird       BRD1818   
4  2019-08-01 00:00:06.593        Bird       BRD1820   
..                     ...         ...           ...   
73 2019-08-01 04:53:48.000        JUMP          JMP2   
74 2019-08-01 04:53:48.000        JUMP          JMP1   
75 2019-08-01 07:04:00.000        JUMP         JMP35   
76 2019-08-01 07:04:00.000        JUMP         JMP34   
77 2019-08-01 00:03:30.973        Lyft         LFT12   

                                         sumdid  tripduration  tripdistance  \
0                                 Powered318477      3.226600     790.68244   
1                                  PoweredYSEGE      7.000000    1968.50400   
2                                  PoweredYTT8X      8.000000       0.00000   
3                                  Powered799T4      7.000000       0.00000   
4                                  PoweredJF4AU     12.000000       0.00000   
..                                          ...           ...           ...   
73  Powereda4712099-5e55-5332-996c-d6e5c910535d      3.154917    1320.00000   
74  Poweredb8a3a269-d1ca-571f-9b2f-89b7399b5537     18.440650    7920.00000   
75  Poweredf077a919-d569-5e70-8ca7-71d179ffacf9    142.345610   20433.60000   
76  Powered784fcc92-bd1a-5f6a-8314-b02aa21b4bfa    209.409650   14889.60000   
77                                Powered599249      4.203000     790.68244   

     startdate        starttime     enddate          endtime  startlatitude  \
0   2019-07-31  23:57:48.396666  2019-08-01  00:01:01.993333      36.157490   
1   2019-07-31  23:54:46.096666  2019-08-01  00:01:29.436666      36.151100   
2   2019-07-31  23:52:40.086666  2019-08-01  00:00:41.923333      36.146100   
3   2019-07-31  23:53:17.616666  2019-08-01  00:00:27.170000      36.158100   
4   2019-07-31  23:48:24.240000  2019-08-01  00:00:16.660000      36.161200   
..         ...              ...         ...              ...            ...   
73  2019-07-31         23:57:19  2019-08-01         00:00:28      36.153015   
74  2019-07-31         23:41:52  2019-08-01         00:00:18      36.153687   
75  2019-07-31         23:26:15  2019-08-01         01:48:35      36.155735   
76  2019-07-31         22:19:08  2019-08-01         01:48:33      36.160050   
77  2019-07-31  23:59:18.720000  2019-08-01  00:03:30.900000      36.164790   

    startlongitude  endlatitude  endlongitude  \
0       -86.777140    36.157950    -86.776030   
1       -86.783300    36.152000    -86.791100   
2       -86.799500    36.151800    -86.797300   
3       -86.769200    36.158100    -86.769200   
4       -86.770900    36.162500    -86.774300   
..             ...          ...           ...   
73      -86.783585    36.152843    -86.779594   
74      -86.784580    36.164158    -86.776900   
75      -86.775185    36.173904    -86.785450   
76      -86.776720    36.163410    -86.782100   
77      -86.779840    36.165800    -86.777830   

                                            triproute               create_dt  
0   [(36.15749, -86.77714), (36.15748, -86.7772), ... 2019-08-02 10:27:50.310  
1   [(36.151103, -86.783327), (36.150872, -86.7831... 2019-08-02 05:30:21.230  
2   [(36.14615, -86.79952), (36.146129, -86.799557... 2019-08-02 05:30:21.340  
3                           [(36.158103, -86.769174)] 2019-08-02 05:30:21.290  
4   [(36.161257, -86.770772), (36.161298, -86.7707... 2019-08-02 05:30:21.373  
..                                                ...                     ...  
73  [('36.153016', '-86.783586'), ('36.153016', '-... 2019-08-02 08:21:26.773  
74  [('36.153685', '-86.784578'), ('36.153685', '-... 2019-08-02 08:21:26.803  
75  [('36.155736', '-86.775181'), ('36.155531', '-... 2019-08-02 08:24:21.967  
76  [('36.160051', '-86.77672'), ('36.160051', '-8... 

Both tables contain 3 months of data and both begin on May 1st, but the trips table ends a day after the scooters table. Looking at why, the trips all began before midnight on July 31st.

In [6]:
long_trips = '''
SELECT 
    companyname,
    COUNT(*)
FROM trips
WHERE tripduration > 1440
GROUP BY companyname;
'''

with engine.connect() as connection:
    long = pd.read_sql(text(long_trips), con = connection)

long_trips

'\nSELECT \n    companyname,\n    COUNT(*)\nFROM trips\nWHERE tripduration > 1440\nGROUP BY companyname;\n'

In [7]:
short_trips = '''
SELECT
    companyname,
    COUNT(*)
FROM trips
WHERE tripduration < 1
    AND tripdistance <= 0
GROUP BY companyname;
'''

with engine.connect() as connection:
    short = pd.read_sql(text(short_trips), con = connection)

short

companyname  count
0        Bird   3928
1        Lime    377
2        Lyft   3405

This data was supposed to have been cleaned before being submitted to the city, which includes stripping out all trips shorter than a minute or longer than 24 hours.

There are ~7,000 trips longer than 24 hours and over 9,000 trips shorter than one minute.

In [8]:
available_scooters = '''
SELECT
    companyname company,
    COUNT(DISTINCT sumdid) scooters
FROM scooters
GROUP BY companyname
'''

with engine.connect() as connection:
    available = pd.read_sql(text(available_scooters), con = connection)

active_scooters = '''
SELECT
    companyname company,
    COUNT(DISTINCT sumdid) scooters
FROM trips
GROUP BY companyname
'''

with engine.connect() as connection:
    active = pd.read_sql(text(active_scooters), con = connection)

available

company  scooters
0    Bird      3860
1    Bolt       360
2  Gotcha       224
3    Jump      1210
4    Lime      1824
5    Lyft      1735
6    Spin       805

In [9]:
active

company  scooters
0           Bird      3766
1  Bolt Mobility       356
2         Gotcha       166
3           JUMP       450
4           Lime      1788
5           Lyft      1725
6           SPIN       754

In [10]:
availability = '''
SELECT
    companyname company,
    COUNT(DISTINCT sumdid) total_scooters
FROM scooters
WHERE sumdid NOT IN
    (SELECT
        DISTINCT sumdid
    FROM trips)
GROUP BY companyname;
'''

with engine.connect() as connection:
    unavailable = pd.read_sql(text(availability), con = connection)
    
unavailable

company  total_scooters
0    Bird             105
1    Bolt               4
2  Gotcha              58
3    Jump             761
4    Lime              73
5    Lyft              12
6    Spin              51

In [11]:
daily_use = '''
SELECT 
    DISTINCT sumdid,
    companyname company,
    ROUND(AVG(COUNT(sumdid)) OVER (PARTITION BY sumdid, DATE(pubtimestamp)),2) avg_daily_usage
FROM trips
GROUP BY sumdid, company, pubtimestamp
'''

with engine.connect() as connection:
    usage = pd.read_sql(text(daily_use), con = connection)

multiple_uses = usage[usage['avg_daily_usage'] > 1.00]
multiple_uses

sumdid        company  \
7      Powered-0555d61d-1c59-4219-edf3-0f00e286629a  Bolt Mobility   
23     Powered-0f344f74-7253-b66b-31b5-ea11fc0434a9  Bolt Mobility   
31     Powered-17bf864f-72d0-d3a2-9100-ba94bcd9adfb  Bolt Mobility   
41     Powered-1d88938b-8196-bd85-c06a-5be82c9a270b  Bolt Mobility   
42     Powered-1d88938b-8196-bd85-c06a-5be82c9a270b  Bolt Mobility   
...                                             ...            ...   
10805                                  PoweredZR8UI           Bird   
10819                          PoweredZTM25VZMQ7JXP           Lime   
10823                                  PoweredZU51C           Bird   
10852                          PoweredZZNTWU4QI7MLI           Lime   
10855                          PoweredZZQOO54WD4AJT           Lime   

       avg_daily_usage  
7                 1.20  
23                1.14  
31                1.25  
41                1.50  
42                2.00  
...                ...  
10805             1.50  
10819             1.43  
10823             1.20  
10852             1.20  
10855             1.43  

[1884 rows x 3 columns]

Filter trips table:
- Remove trips under a minute
- Remove all zero-distance trips
- Remove unreasonable long trips
    - These scooters have an average top speed of 15 mph, and generally have a range of about 50 miles.
    - That said, the batteries should last on average ~3.5 hours

In [12]:
trips_clean = '''
SELECT *
FROM trips
WHERE tripduration > 1.0
    AND tripduration < 200.0
    AND tripdistance > 0
    AND tripdistance < 264000
'''

with engine.connect() as connection:
    trips = pd.read_sql(text(trips_clean), con = connection)
    
trips.describe()

trips.head()

pubtimestamp companyname triprecordnum                sumdid  \
0 2019-05-11 16:02:02.567        Lime       LIM1090  PoweredAMLAI4DWT2SXM   
1 2019-05-11 16:02:02.570        Lime       LIM1098  PoweredOYKGTYGE4MKYC   
2 2019-05-11 16:02:29.263        Lyft        LFT612         Powered233573   
3 2019-05-11 16:02:02.570        Lime       LIM1094  Powered3WD63MPJQERKZ   
4 2019-05-11 16:02:02.570        Lime       LIM1096  Powered7VFN2ISITXNKP   

   tripduration  tripdistance   startdate        starttime     enddate  \
0     14.016667      599.2368  2019-05-11         15:43:30  2019-05-11   
1     14.833333      945.4896  2019-05-11         15:44:42  2019-05-11   
2     10.487317     3641.7324  2019-05-11  15:51:59.926666  2019-05-11   
3      4.650000      116.7384  2019-05-11         15:53:15  2019-05-11   
4      3.800000       43.8912  2019-05-11         15:54:45  2019-05-11   

           endtime  startlatitude  startlongitude  endlatitude  endlongitude  \
0         15:57:31      36.157231      -86.774504    36.143076    -86.768277   
1         15:59:32      36.148278      -86.805935    36.157998    -86.776818   
2  16:02:29.166666      36.175210      -86.788040    36.170010    -86.787600   
3         15:57:54      36.136769      -86.799551    36.136995    -86.788301   
4         15:58:33      36.138073      -86.800774    36.137854    -86.800883   

                                           triproute               create_dt  
0  [(36.15709, -86.77449), (36.157109999999996, -... 2019-05-12 07:59:50.383  
1  [(36.14831, -86.80588), (36.14858, -86.80537),... 2019-05-12 07:59:50.643  
2  [(36.17521, -86.78804), (36.17524, -86.78809),... 2019-05-12 10:40:14.753  
3  [(36.13682, -86.79955), (36.13675, -86.79955),... 2019-05-12 07:59:50.513  
4  [(36.13807, -86.80077), (36.13785, -86.80077),... 2019-05-12 07:59:50.580

**MAPPING**

In [13]:
zipcodes = gpd.read_file('../data/zipcodes.geojson')

zipcodes = zipcodes[['zip', 'po_name', 'geometry']]
# Create separate tables for start & end location data
scooter_start = trips[['companyname', 'triprecordnum', 'sumdid', 'startlongitude', 'startlatitude']]
scooter_start_geo = gpd.GeoDataFrame(scooter_start, crs= zipcodes.crs, geometry=gpd.points_from_xy(scooter_start.startlongitude, scooter_start.startlatitude))
scooter_end = trips[['companyname', 'triprecordnum', 'sumdid', 'endlongitude', 'endlatitude']]
scooter_end_geo = gpd.GeoDataFrame(scooter_end, crs= zipcodes.crs, geometry=gpd.points_from_xy(scooter_end.endlongitude, scooter_end.endlatitude))

In [14]:
# Combine start and end locations by zip code
starting_zip = gpd.sjoin(scooter_start_geo, zipcodes, predicate='within')
ending_zip = gpd.sjoin(scooter_end_geo, zipcodes, predicate='within')

In [15]:
starting_zip.head()

companyname triprecordnum                sumdid  startlongitude  \
0         Lime       LIM1090  PoweredAMLAI4DWT2SXM      -86.774504   
16        Lyft        LFT613         Powered661244      -86.777530   
25        Bird        BRD578          PoweredDPCL6      -86.774400   
55        Bird        BRD619          PoweredUL4Y4      -86.779900   
64        Lime       LIM1120  PoweredUWLZ32JTQZWD4      -86.774900   

    startlatitude                    geometry  index_right    zip    po_name  
0       36.157231  POINT (-86.77450 36.15723)           36  37201  NASHVILLE  
16      36.159660  POINT (-86.77753 36.15966)           36  37201  NASHVILLE  
25      36.157000  POINT (-86.77440 36.15700)           36  37201  NASHVILLE  
55      36.165200  POINT (-86.77990 36.16520)           36  37201  NASHVILLE  
64      36.157580  POINT (-86.77490 36.15758)           36  37201  NASHVILLE

In [16]:
ending_zip.head()

companyname triprecordnum                sumdid  endlongitude  endlatitude  \
0        Lime       LIM1090  PoweredAMLAI4DWT2SXM    -86.768277    36.143076   
1        Lime       LIM1098  PoweredOYKGTYGE4MKYC    -86.776818    36.157998   
3        Lime       LIM1094  Powered3WD63MPJQERKZ    -86.788301    36.136995   
6        Lime       LIM1093  PoweredU6FP6ZXZ6JRTC    -86.781084    36.159457   
7        Lime       LIM1097  Powered2OA72HN6HFAJH    -86.797570    36.154829   

                     geometry  index_right    zip    po_name  
0  POINT (-86.76828 36.14308)           35  37203  NASHVILLE  
1  POINT (-86.77682 36.15800)           35  37203  NASHVILLE  
3  POINT (-86.78830 36.13699)           35  37203  NASHVILLE  
6  POINT (-86.78108 36.15946)           35  37203  NASHVILLE  
7  POINT (-86.79757 36.15483)           35  37203  NASHVILLE

Now let's look at where scooter trips are taking people. Counting up the number of trips that began in each zip code and comparing them with the number of trips ending in each zip code, we can see which areas people are more likely to come from and go to.

In [17]:
start_by_zip = pd.DataFrame(starting_zip['zip'].value_counts())
end_by_zip = pd.DataFrame(ending_zip['zip'].value_counts())
delta = lambda start_by_zip, end_by_zip : end_by_zip - start_by_zip

In [18]:
destinations = delta(start_by_zip,end_by_zip).sort_values(by='count', ascending=False)
destinations.head(3)

count
zip          
37201  1804.0
37207  1022.0
37209   910.0

In [19]:
origins = delta(start_by_zip,end_by_zip).sort_values(by='count', ascending=True)
origins.head(3)

count
zip          
37204 -4072.0
37219 -2263.0
37203 -1066.0

More trips ended in 37201, 37207, and 37209 than started there, while more trips began in 37204, 37219, and 37203 than ended there. Now let's look into trips to/from these zip areas.

In [20]:
origin_destination_trips= trips[['companyname', 'triprecordnum', 'sumdid', 'startlongitude', 'startlatitude', 'endlongitude', 'endlatitude']]
origin_trips_geo = gpd.GeoDataFrame(origin_destination_trips, crs= zipcodes.crs, geometry=gpd.points_from_xy(origin_destination_trips.endlongitude, origin_destination_trips.endlatitude))
dest_trips_geo = gpd.GeoDataFrame(origin_destination_trips, crs= zipcodes.crs, geometry=gpd.points_from_xy(origin_destination_trips.startlongitude, origin_destination_trips.startlatitude))

In [21]:
origin_zip = gpd.sjoin(origin_trips_geo, zipcodes, predicate='within')
destination_zip = gpd.sjoin(dest_trips_geo, zipcodes, predicate='within')

In [29]:
origin_zip['zip'].value_counts()

zip
37203    207462
37201     84478
37219     35431
37212     30705
37204     22171
37208     20219
37206     18337
37213     16860
37210     11227
37240      4095
37207      3258
37209      2537
37232      2227
37215       904
37216       779
37228       670
37205       575
37211       214
37217        81
37214        69
37218        63
37013        43
37115        33
37221        18
37220         4
37027         2
37072         2
37138         2
37076         1
37189         1
Name: count, dtype: int64

In [30]:
destination_zip['zip'].value_counts()

zip
37203    208528
37201     82674
37219     37694
37212     30348
37204     26243
37208     19399
37206     17604
37213     17448
37210     10756
37240      3777
37232      2364
37207      2236
37209      1627
37215       453
37228       424
37216       369
37205       302
37211        88
37217        47
37214        39
37013        39
37218        23
37115        15
37221        10
37138         2
37220         2
37072         1
Name: count, dtype: int64

In [24]:
origin_37204 = gpd.GeoDataFrame(origin_zip[origin_zip['zip']=='37204'])
origin_37219 = gpd.GeoDataFrame(origin_zip[origin_zip['zip']=='37219'])
origin_37203 = gpd.GeoDataFrame(origin_zip[origin_zip['zip']=='37203'])

In [25]:
origin_37204.head()

companyname triprecordnum                sumdid  startlongitude  \
61        Lime       LIM1622  PoweredPUE3JGPNOICN6      -86.790849   
64        Lime       LIM1120  PoweredUWLZ32JTQZWD4      -86.774900   
66        Lime       LIM1125  Powered25UE3EUVBN6RU      -86.774795   
71        Lime       LIM1124  PoweredGZYIKLJC3TZ2P      -86.774904   
72        Lime       LIM1127  Powered7BYCHGP3RAHWA      -86.774705   

    startlatitude  endlongitude  endlatitude                    geometry  \
61      36.121388    -86.798531    36.106717  POINT (-86.79853 36.10672)   
64      36.157580    -86.789262    36.126519  POINT (-86.78926 36.12652)   
66      36.157605    -86.789206    36.126540  POINT (-86.78921 36.12654)   
71      36.157600    -86.789186    36.126598  POINT (-86.78919 36.12660)   
72      36.157502    -86.789163    36.126583  POINT (-86.78916 36.12658)   

    index_right    zip    po_name  
61            2  37204  NASHVILLE  
64            2  37204  NASHVILLE  
66            2  37204  NASHVILLE  
71            2  37204  NASHVILLE  
72            2  37204  NASHVILLE

In [26]:
destination_37201 = gpd.GeoDataFrame(destination_zip[destination_zip['zip']=='37201'])
destination_37207 = gpd.GeoDataFrame(destination_zip[destination_zip['zip']=='37207'])
destination_37209 = gpd.GeoDataFrame(destination_zip[destination_zip['zip']=='37209'])

In [27]:
destination_37201.head()

companyname triprecordnum                sumdid  startlongitude  \
0         Lime       LIM1090  PoweredAMLAI4DWT2SXM      -86.774504   
16        Lyft        LFT613         Powered661244      -86.777530   
25        Bird        BRD578          PoweredDPCL6      -86.774400   
55        Bird        BRD619          PoweredUL4Y4      -86.779900   
64        Lime       LIM1120  PoweredUWLZ32JTQZWD4      -86.774900   

    startlatitude  endlongitude  endlatitude                    geometry  \
0       36.157231    -86.768277    36.143076  POINT (-86.77450 36.15723)   
16      36.159660    -86.783970    36.158240  POINT (-86.77753 36.15966)   
25      36.157000    -86.766800    36.144800  POINT (-86.77440 36.15700)   
55      36.165200    -86.778400    36.157900  POINT (-86.77990 36.16520)   
64      36.157580    -86.789262    36.126519  POINT (-86.77490 36.15758)   

    index_right    zip    po_name  
0            36  37201  NASHVILLE  
16           36  37201  NASHVILLE  
25           36  37201  NASHVILLE  
55           36  37201  NASHVILLE  
64           36  37201  NASHVILLE

In [28]:
# Get centroid of Davidson County
county = gpd.read_file('../data/Davidson County Border (GIS).geojson')
county.geometry.centroid
# Use ID #0
center = county.geometry.centroid[0]
map_center = [center.y, center.x]

scooters_37201 = folium.Map(location = map_center, zoom_start = 8)
# Add marker cluster
marker_cluster = MarkerCluster().add_to(scooters_37201)
# Add zipcodes to map
folium.GeoJson(zipcodes).add_to(scooters_37201)
# Use a for loop to add projects
for row_index, row_values in scooters_37201.iterrows():
    loc = [row_values['startlatitude'], row_values['startlongitude']]
    pop = str(row_values['companyname'])
    icon=folium.Icon(color="blue",icon="exclamation-triangle", prefix='fa')
    
    marker = folium.Marker(
        location = loc, 
        popup = pop,
   icon = icon) 
    
    marker.add_to(marker_cluster)

scooters_37201.save('../maps/scooters_37201.html')
# Display the map
scooters_37201

C:\Users\chabi\AppData\Local\Temp\ipykernel_18208\1488178287.py:3: UserWarning: Geometry is in a geographic CRS. Results from 'centroid' are likely incorrect. Use 'GeoSeries.to_crs()' to re-project geometries to a projected CRS before this operation.

  county.geometry.centroid
C:\Users\chabi\AppData\Local\Temp\ipykernel_18208\1488178287.py:5: UserWarning: Geometry is in a geographic CRS. Results from 'centroid' are likely incorrect. Use 'GeoSeries.to_crs()' to re-project geometries to a projected CRS before this operation.

  center = county.geometry.centroid[0]


AttributeError: 'Map' object has no attribute 'iterrows'

In [ ]:
# Get centroid of Davidson County
county = gpd.read_file('../data/Davidson County Border (GIS).geojson')
county.geometry.centroid
# Use ID #0
center = county.geometry.centroid[0]
map_center = [center.y, center.x]

scooters_37207 = folium.Map(location = map_center, zoom_start = 8)
# Add marker cluster
marker_cluster = MarkerCluster().add_to(scooters_37207)
# Add zipcodes to map
folium.GeoJson(zipcodes).add_to(scooters_37207)
# Use a for loop to add projects
for row_index, row_values in list(scooters_37201.iterrows()):
    loc = [row_values['startlatitude'], row_values['startlongitude']]
    pop = str(row_values['companyname'])
    icon=folium.Icon(color="blue",icon="exclamation-triangle", prefix='fa')
    
    marker = folium.Marker(
        location = loc, 
        popup = pop,
   icon = icon) 
    
    marker.add_to(marker_cluster)

scooters_37207.save('../maps/scooters_37207.html')
# Display the map
scooters_37207